pyro/infer/autoguide/guides.py

```diff
@@ -217,7 +236,7 @@ def __call__(self, *args, **kwargs):
         return {} if result is None else result
 
 
-class AutoDelta(AutoGuide):
+class AutoDelta(AutoGuide, ConstrainedModule):
```

Comment:
It's is probably safer to avoid multiple inheritance if possible. One way is to avoid letting `AutoGuide` be an `nn.Module`:
```py
class AutoGuide: ...
class AutoGuideList(AutoGuide, nn.Module): ...
class AutoDelta(AutoGuide, ConstrainedModule): ...
class AutoContinuous(AutoGuide, nn.Module): ...
class Auto*Normal(AutoContinuous): ...
```

Reply:
The AutoDelta's `__mro__` looks fine, i.e. `ConstrainedModule` precedes `nn.Module`. Are there other considerations? 

```
<class 'pyro.infer.autoguide.guides.AutoDelta'>, 
<class 'pyro.infer.autoguide.guides.AutoGuide'>, 
<class 'pyro.params.constrained_parameter.ConstrainedModule'>, 
<class 'torch.nn.modules.module.Module'>, 
<class 'object'>)
```

We can also just use the hierarchy above - the only downside that I can see is that we cannot call module methods like `named_parameters` unless we know which autoguide it is. Doesn't seem like a big deal.

Reply:
OK, if you're happy with `AutoDelta`'s `__mro__`, I'm fine with `AutoGuide(nn.Module)`.

Reply:
I just followed your suggestion which would keep the parent classes intact for other custom autoguides, and was a simple fix. In any case, those classes like the ones passed to `autocallable` will need to modified for them to be practically usable as `nn.Module`s.